### PR TITLE
Terraform improvements

### PR DESCRIPTION
**Pull request by syedhassaanahmed**
This PR aims to improve the Terraform templates in the following manner
- A dev container is provided which helps setup the dev environment for .NET and Terraform
- The CI pipeline has an additional job which initializes and lints the Terraform template
- Azure resource names conform to the [CAF recommended abbreviations](https://docs.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-abbreviations)
- Azure Terraform provider was bumped to v3.2. This introduced some changes to the App Service and App Service Plan resources
- The App Service uses Managed Identity to authenticate to the ACR
- `null_resources` have `destroy` scripts
- DPS enrollment script is made idempotent to avoid errors in case previous deployment was unsuccessful. This script also explicitly sets `use_dynamic_install=yes_without_prompt` to avoid install prompts during `terraform apply`
- Default username for the IoT Edge VM has been changed to `azureuser`
- Readme contains a note about installing .NET Runtime 6 which is used by the Terraform template to execute API Facade Console App. The API Facade `null_resource` needs to be executed before the DPS enrollment, hence this block was moved inside the Key Vault module.

### CODE REVIEW

File: terraform/modules/iot-hub/main.tf [hunk 1]
```diff
@@ -68,15 +68,22 @@ resource "azurerm_iothub_dps" "iot_dps" {
...
-    command     = "az iot dps certificate create --certificate-name ${var.issuing_ca} --dps-name ${azurerm_iothub_dps.iot_dps.name} --path ${var.issuing_ca}.cer --resource-group ${var.resource_group_name} --verified true"
-  }
+    command     = <<EOF
+      set -Eeuo pipefail
 
-  provisioner "local-exec" {
-    command = "az iot dps enrollment-group create -g ${var.resource_group_name} --dps-name ${azurerm_iothub_dps.iot_dps.name}  --enrollment-id ${var.resource_prefix}-enrollmentgroup --edge-enabled true --ca-name ${var.issuing_ca}"
+      CERT_NAME=$(az iot dps certificate list -g ${var.resource_group_name} --dps-name ${azurerm_iothub_dps.iot_dps.name} --query "value[?name=='${var.issuing_ca}'].name" -o tsv)
+
+      if [ -z "$CERT_NAME" ]
```
Comment (Ioana37) [inline terraform/modules/iot-hub/main.tf:85]: Please also make sure to delete the local certificate when destroying the resources.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -22,10 +22,29 @@ jobs:
...
+    - name: Checkout
+      uses: actions/checkout@v3
+
+    - name: Setup Terraform
+      uses: hashicorp/setup-terraform@v1
+
+    - name: Terraform Init
+      run: terraform init -input=false
+
+    - name: Terraform Format
```
Comment (Ioana37) [inline .github/workflows/ci.yml:49]: Can you also add a plan step? That would ensure that the Terraform content is fully valid.

File: terraform/modules/acr/main.tf [hunk 3]
```diff
@@ -8,6 +8,6 @@ resource "azurerm_container_registry" "acr" {
 
 resource "null_resource" "push-docker" {
   provisioner "local-exec" {
-    command = "az acr build --image sample/estserver:v2 --registry ${azurerm_container_registry.acr.name} https://github.com/vslepakov/keyvault-ca.git --file ./././KeyVaultCA.Web/Dockerfile"
+    command = "az acr build -r ${azurerm_container_registry.acr.name} --image sample/estserver:v2 https://github.com/vslepakov/keyvault-ca.git -f ./././KeyVaultCA.Web/Dockerfile"
```
Comment (Ioana37) [inline terraform/modules/acr/main.tf]: Could we also use a short parameter for the '--image'?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (syedhassaanahmed) [reply .github/workflows/ci.yml:49]: Plan step unfortunately requires authentication to Terraform. Should we do it already or as part of the CD pipeline?

File: terraform/modules/iot-hub/main.tf (hunk 1 above)
Comment (syedhassaanahmed) [reply terraform/modules/iot-hub/main.tf:85]: Done

File: terraform/modules/acr/main.tf (hunk 3 above)
Comment (syedhassaanahmed) [reply terraform/modules/acr/main.tf]: Done

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Ioana37) [reply .github/workflows/ci.yml:49]: I see, then makes sense to leave it for the CD.

File: terraform/modules/acr/output.tf [hunk 4]
```diff
@@ -1,17 +1,7 @@
...
-  value = azurerm_container_registry.acr.name
+output "acr_id" {
+  value = azurerm_container_registry.acr.id
 }
 
 output "acr_login_server" {
   value = azurerm_container_registry.acr.login_server
-}
-
-output "acr_admin_username" {
```
Comment (machteldbogels) [inline terraform/modules/acr/output.tf]: I'll need to test if the Edge VM can also make use of MI for authenticating to ACR in the private setup (currently I'm still using username/password for pulling the edgeAgent image) - so we might need to re-introduce these variables later but that should be fine.

Comment (syedhassaanahmed) [reply terraform/modules/acr/output.tf]: @machteldbogels Did you mean Edge VM using MI to authenticate to ACR for the IoT Edge modules in the deployment manifest?

Comment (machteldbogels) [reply terraform/modules/acr/output.tf]: @syedhassaanahmed yes exactly. so far it seems that adding a role assignment for the VM and then removing username/password does not allow the pull automagically. according to these [docs](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-authentication-managed-identity#use-the-identity-to-access-the-registry-1) I'll need to use the az cli to ensure the system identity is being used, but we wouldn't want to install this on the vm (I assume)

Comment (syedhassaanahmed) [reply terraform/modules/acr/output.tf]: Got it. In that case I'll revert back to outputting the ACR username and password as managed identity won't work on a real edge device anyway.

Comment (syedhassaanahmed) [reply terraform/modules/acr/output.tf]: Done

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.191.0/containers/dotnetcore
```
Comment (machteldbogels) [inline .devcontainer/devcontainer.json]: link seems to be broken

Comment (syedhassaanahmed) [reply .devcontainer/devcontainer.json]: Good catch. Fixed.

**Comment by Ioana37** [comment]
I've redeployed all the Terraform resources using the dev container and it works smoothly, so I will merge the PR.
